sources/host-ctr/cmd/host-ctr/pid_namespaces.go

```diff
@@ -0,0 +1,184 @@
+	manifestPath  string = "/etc/host-containers/info/pid-namespaces.json"
+)
+
+func getTaskPidNs(task *containerd.Task) (int64, error) {
+	pid := (*task).Pid()
+	pidNsFile, err := os.Open(fmt.Sprintf("/proc/%d/ns/pid", pid))
+	if err != nil {
+		return -1, fmt.Errorf("failed to open %d PID namespace file", pid)
+	}
+	pidNsFd := int(pidNsFile.Fd())
+
+	_, err = unix.IoctlGetInt(pidNsFd, nsGetParentNs)
+	if err == syscall.EPERM {
+		// If we can't access the parent PID namespace, this means the current container task is sharing the root PID
+		// namespace. We don't want to track the root PID namespace in the manifest, so we return -1.
+		return -1, nil
+	} else if err != nil {
+		return -1, errors.Wrap(err, fmt.Sprintf("ioctl failed to 'NS_GET_PARENTNS' for '%s'", pidNsFile.Name()))
+	}
+
+	// The PID namespace is represented by the inode number of the task's PID namespace file
+	var sb syscall.Stat_t
+	if err := syscall.Fstat(pidNsFd, &sb); err != nil {
+		return -1, errors.Wrap(err, "failed to stat task PID namespace fd")
+	}
+
+	return int64(sb.Ino), nil
+}
+
+// HostCtrPidNs represents the list of PID namespaces
+type HostCtrPidNs struct {
+	PidNamespaces []int64 `json:"pid-namespaces,omitempty"`
+}
+
+func addPidNsToManifest(pidNs int64) error {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	locked, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	if !locked {
+		return errors.New("failed to acquire lock for writing to PID NS manifest")
+	}
+	defer func() {
+		_ = fileLock.Unlock()
+	}()
+
+	// Add task PID NS to manifest
+	if err := os.MkdirAll(filepath.Dir(manifestPath), 0666); err != nil {
+		return errors.Wrap(err, "failed to create PID NS manifest directory")
+	}
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
+		return errors.Wrap(err, "failed to read PID NS manifest file")
+	}
+	var (
+		tmpPidNsList     HostCtrPidNs
+		hostCtrPidNsList HostCtrPidNs
+	)
+	if len(raw) != 0 {
+		err = json.Unmarshal(raw, &tmpPidNsList)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal PID NS manifest")
+		}
+		for _, ele := range tmpPidNsList.PidNamespaces {
+			// If it already exists in the manifest list for whatever reason, we can just return without doing anything.
+			if ele == pidNs {
+				return nil
+			}
+		}
+	}
+	hostCtrPidNsList.PidNamespaces = append(tmpPidNsList.PidNamespaces, pidNs)
+	newData, err := json.MarshalIndent(hostCtrPidNsList, "", "  ")
```

Comment:
I chose to remove it for now since it's mostly meant for machines and there's not much going on here anyways to make it hard to read. We can always revisit this when we decide to add more to it.